### ISSUE TITLE
OBB - Ground truth labels getting distored and inaccurate in train batch images (with incorrect orientations and offsets)

### ISSUE THREAD

**Issue by farhanrw** [ultralytics/ultralytics#10181]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

Hi, firstly thanks to the whole team for the amazing work in developing yolov8 (especially with OBB). I am still seeing issues with OBB with and without data augmentation. I am attaching two images as an example. When using OBB, I can constantly see the ground truth overlays are sometimes not oriented and sometimes not fitting around all the edges. By cases I mean from what I see in train_batch0,1 and 2 inside the train folder.

To illustrate this problem clearly, I created a dummy dataset of 4-5 images with a colorful rectangle sitting at the center and labeled the images. I labeled these myself and exported them using Roboflow and Yolov8-obb. As you can see from the images, the square shapes are not always showing correctly oriented bounding boxes. My dataset just has 4-5 images and I have manually hand-labeled them correctly. This does not happen when I use the regular detection model without OBB. Using OBB and additional rotation augmentation distorts the boxes, especially at the edges.

These attached images are not predictions, but rather train_batch images.

![image](https://github.com/ultralytics/ultralytics/assets/43019847/c204689e-c1e0-4dc7-9da2-f37b25a7d72e)
![image](https://github.com/ultralytics/ultralytics/assets/43019847/38a0c3ad-944d-4e5c-a966-aece302b02c4)

A few more with a larger resolution. You can also see how some of these boxes are not tight and seem to have an offset (bottom right one as an example)

![image](https://github.com/ultralytics/ultralytics/assets/43019847/60b577b9-b545-4e1a-a440-e220d82cfea7)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
Hello! Thanks for sharing your observations with the OBB (Oriented Bounding Box) functionality in YOLOv8, and for providing detailed examples from your experiment. 😊 Your contribution is invaluable to refining and improving the model.

From your description, it sounds like the ground truth overlays for OBBs are not aligning as expected, particularly with orientation and fitting issues, notably after applying rotation augmentation. This could be influenced by how augmentation transformations are applied or interpreted for OBBs compared to standard bounding boxes.

To address this, could you please try adjusting the augmentation parameters in your training configuration? Specifically, you might want to experiment with modifying or temporarily disabling rotation augmentation to assess its direct impact on OBB distortion.

For example, in your dataset's YAML configuration, you can adjust the `degrees` parameter under `augmentations` to a lower value or set it to 0 to disable rotation:
```yaml
augment:
  degrees: 0  # Adjust or set to 0 to disable rotation augmentation
```
This step will help isolate if the issue stems directly from how the model handles rotation augmentations with OBBs.

Additionally, it's worth ensuring that your dataset annotations for OBBs are in the expected format YOLOv8 requires. A minor discrepancy in annotation format could lead to unexpected behaviors.

Lastly, if the issue persists, it might be beneficial to share a snippet of your dataset (if possible) and the specific configuration settings you're using. That information could help the community or the dev team to replicate the scenario more accurately and provide further assistance.

Thank you for your detailed report, and we look forward to assisting you in resolving this matter! 🚀

**Comment by glenn-jocher** [comment]
@farhanrw can you share your dataset in a zip here in this issue please to allow us to reproduce?

@Laughing-q user here is saying that rotation augmentation is not working well with OBB, can you check this out?

**Comment by farhanrw** [comment]
@glenn-jocher @Laughing-q sorry for the late reply. Yes sure, this is a dummy dataset that I created and can share it.
[Test.v2i.yolov8-obb.zip](https://github.com/ultralytics/ultralytics/files/15067289/Test.v2i.yolov8-obb.zip)

**Comment by farhanrw** [comment]
Also it might just not be related to the rotation augmentation. If you set the degrees to 0 but still have mosaic on then you can also notice that the mosaiced images where there are bounding boxes at the edge of the images, the boxes seem to have an offset, as in it does not go over the edge of the object accurately. Again this can be reproduced but using 0 deg rotation augmentation and using mosaic.

![image](https://github.com/ultralytics/ultralytics/assets/43019847/971d595c-d0b0-437c-ac8b-5ad846fbee69)

See how in the bottom two images the one on the left does not cover the partial object properly, and how the bottom right does not have the bounding box?

**Comment by glenn-jocher** [comment]
Thanks for the additional observations, it’s incredibly helpful! 🙏 It seems like the issue with offsets around object edges, particularly near image boundaries, persists even without rotation augmentation and can indeed be more prominent with mosaic augmentation enabled. This could be related to how mosaic images are constructed and how bounding boxes are recalculated for the stitched together scenes.

For the mosaic augmentation specifically, one approach to mitigate this could be to refine how bounding boxes are adjusted during the mosaic assembly process. Ensuring that bounding boxes remain tightly fitted around objects, especially when they're near the edges of the contributing images, is crucial. 

Here's a snippet that shows a basic outline for handling boundary conditions in mosaic augmentation, which could be fine-tuned further based on specific dataset characteristics:

```python
if bbi_intersects_image_boundary:
    # Adjust bounding box to fit within the boundaries of the mosaic image
    bbi = adjust_bounding_box(bbi, mosaic_image_boundary)
```

Certainly, this is a simplified overview, and the actual implementation might need a deep dive into the coordinate transformations and checks involved in creating mosaic datasets.

Your insight on observing these issues across different augmentation settings is invaluable. We will delve into this to determine optimal handling for such cases, enhancing data augmentation methods, particularly for objects at image boundaries, ensuring accuracy in label placements.

Your shared examples clearly illustrate the issue, and we’ll be working on ways to address it. If you encounter any more peculiarities or have suggestions, please keep sharing! Your contributions help make YOLOv8 even better. 🚀

**Comment by glenn-jocher** [comment]
@farhanrw thanks for the additional observations, it’s incredibly helpful! 🙏 Your insights on the potential impact of mosaic augmentation, even with rotation disabled (degrees set to 0), shine a light on a broader aspect of how augmented bounding boxes are handled. It seems that objects at the edges of mosaiced images might not get their bounding boxes accurately fitted, impacting the precision of OBB.

Investigating the behavior with mosaic augmentation, especially at the image edges, is definitely on our radar now. The examples you've provided clearly illustrate the problem. It's possible that the algorithm’s handling of edge cases (literally!) in mosaic-augmented images needs fine-tuning.

For now, one immediate step you could take is to experiment with the `mosaic` augment parameter by either adjusting its probability or disabling it temporarily to see if that affects the accuracy of the bounding box placement. For example, to disable mosaic augmentation, you could adjust your dataset YAML as follows:

```yaml
# In your dataset.yaml
mosaic: 0  # Set to 0 to disable mosaic augmentation
```

We'll delve into this issue and explore how we can enhance the model's handling of such cases. If you notice any other peculiar behaviors or have further insights, please keep sharing. Every bit of detailed feedback is a great asset in improving YOLOv8! 🚀

**Comment by farhanrw** [comment]
Thanks for the reply. This issue still persists with no mosaic augmentation, i.e. with mosaic: 0.

**Comment by glenn-jocher** [comment]
Hey there! Thanks for the update. 😊 If the issue persists even without mosaic augmentation, this gives us a clearer direction to investigate further. In the meantime, could you try tweaking other augmentation parameters (like `perspective` and `translate`) just to see if any specific augmentation is influencing the accuracy of OBB? For instance:

```yaml
# Customizing augmentations
augment:
  perspective: 0.0  # Disable perspective changes
  translate: 0.0  # Disable translation
```

This could help isolate the problem while we dig deeper on our end. Thanks for your patience and for helping us pinpoint the issue!

**Comment by farhanrw** [comment]
Hi, has there been any update on this? @Burhan-Q @Laughing-q

**Comment by Yeah-l** [comment]
> @glenn-jocher @farhanrw Guys the final augmented oriented bounding boxes are determined by the algorithm `cv2.minAreaRect()` that we uses in `xyxyxyxy2xywhr` function.
> 
> https://github.com/ultralytics/ultralytics/blob/5a82b511074617b9daa25000e5672c5d56188374/ultralytics/utils/ops.py#L539
> 
> 
> For objects that are cut-off a lot by augmentations, current final boxes(shown above) might be the best the algorithm can do.
> I've already figured a solution to alleviate the issue which showed better visual results in my tests on DOTAv1 dataset, that is to sample more(4 --> 100) coordinates(so that `cv2.minAreaRect()` could get `xywhr` from a more dense candidate) for each oriented bounding boxes to get better final results.
> https://github.com/ultralytics/ultralytics/blob/5a82b511074617b9daa25000e5672c5d56188374/ultralytics/data/dataset.py#L219
> 
> @farhanrw From your samples the solution is not working very well on your dataset, could you upload a few samples here if you don't mind? so that we can try to figure a better solution.

@farhanrw @Laughing-q @glenn-jocher Thank you for your great work! However, I found that this situation would still occur if I set 100 interpolation points on bounding box. I think this reason is caused by the deletion of too many interpolation points when cropping, especially if the cropped bounding box ends up with two edges left on the original pic. Relying on interpolation points on the two borders can easily result in misalignment of the generated minimum external rectangle. Below I have given some pictures from dota datasets to show:
![1](https://github.com/ultralytics/ultralytics/assets/141161807/8824ec5f-de2f-4726-b814-76d1e207ccfc)
![2](https://github.com/ultralytics/ultralytics/assets/141161807/a8c80d3e-1b45-4875-9663-42d8e8301435)

**Comment by glenn-jocher** [comment]
Thank you for your detailed insights and for sharing the visual examples from the DOTA dataset! Your observations about the interpolation points and the cropping effects on bounding boxes are very valuable.

It seems that the issue might indeed be related to how the bounding boxes are recalculated after cropping, especially when only two edges remain. This can lead to inaccuracies in the `cv2.minAreaRect()` function's output.

To further investigate and potentially resolve this, we could consider a few approaches:

1. **Increase Interpolation Points**: As you've already tried increasing the number of interpolation points, it might be beneficial to experiment with even higher values or different sampling strategies to see if it improves the bounding box accuracy.

2. **Custom Post-Processing**: Implement a custom post-processing step to refine the bounding boxes after augmentation. This could involve additional checks and adjustments to ensure the bounding boxes remain accurate.

3. **Alternative Algorithms**: Explore alternative algorithms or methods for calculating the minimum enclosing rectangle that might handle edge cases better.

If you have any additional insights or suggestions, please feel free to share them. Your contributions are greatly appreciated, and together we can work towards improving the OBB functionality for everyone. 🚀

Thank you for your patience and collaboration!

**Comment by Bhavay-2001** [comment]
Hi @glenn-jocher, I found this issue in the TODO task. Is it open to work on?

**Comment by PhilippLenski** [comment]
Hey @glenn-jocher,
Hey,

Thank you very much for your great work. I have now encountered the problem myself and wanted to ask if there is a workaround for it or something similar?
For my master thesis I have to recognize rotated alignment markers and due to the error the detection is unfortunately too inaccurate.

I have played with almost all the parameters and even completely deactivated the augmentation of the data, set batch size to 1 only and did not scale the images. Unfortunately, nothing helped.

Can I switch to a different yolo version or does the problem exist for everyone?

**Comment by shamangary** [comment]
Hi first of all great github!
I think this problem is still there when there is rotation aug!

**Comment by glenn-jocher** [comment]
@shamangary thank you for your feedback! The issue with rotation augmentation and OBB is being investigated, and we are exploring ways to improve its performance. If you have specific examples or tests, feel free to share them to assist with debugging.

**Comment by wvalcke** [comment]
Hello,

I investigated this a little bit. The problem lies in the minAreaRect call of opencv. You cannot trust it for getting the correct orientation when we have done some augmentation, especially when the rotated boxes are at the edges. I took a different approach.

- When the label is loaded i determine from the annotation the cx, cy, w, h and angle of the OBB and keep this info separate with other data like box info, segment, etc...
- When augmentation is being performed (Mosaic, affine transformations (scale, rotate, translate), flipping, ..) i track the changes in the separate OBB info. I change the cx, cy values, change the angle, and when scaling i adapt the w,h of the OBB data. Even if the cx, cy would move completely out of the image boundaries i do not remove it.
- Using the (existing already) segment data we can track what piece of the OBB still lies in the image boundaries.
- Now at the Format transform. Instead of using the segment data and minAreaRect i do the following. As i know the orientation of the OBB (this was tracked during augmentation) i can create a coordinate system aligned with the OBB, and search the min/max values of the segment to get the minimal rectangle around the data. By getting the 4 corner points and expressing those back in the image coordinate system i do have the correct OBB corner points in image coordinate system. Now we can easily use minAreaRect here which will always have the correct angle, and this determines a new width and height, as well a (possible) new cx, cy. And simple we now have the correct OBB.
- This was implemented on top of v8.3.99 was then merged into v 8.3.136. 
- The code is seen as an extension, i tried not to break anything and should be backward compatible. Although it is not yet tested on detect, pose, segment & classify training. 
- For OBB, my tests show that the augmentation is now doing the correct thing looking at the first batch images when training is started.

If people are interested, give it a try (for code see above). I hope the explanation is a bit clear. And if not, i will see this from the comments.

**Comment by glenn-jocher** [comment]
@wvalcke Thank you for sharing your detailed investigation and solution! Your approach of tracking OBB parameters (cx, cy, w, h, angle) separately through the augmentation pipeline is very insightful. 

This is an excellent solution that addresses the core issue - the unreliability of `cv2.minAreaRect()` for determining orientation after augmentations, especially at image edges. By maintaining the original OBB orientation and transforming it properly through each augmentation step, you've effectively bypassed the limitations of the current implementation.

Your implementation sounds compatible with our architecture and could be integrated into YOLO11's OBB handling. The coordinate system transformation approach to find the minimal rectangle around segments is particularly clever.

Would you be interested in submitting a PR with this solution to improve our OBB implementation? Your work could significantly enhance the accuracy of oriented bounding boxes training for all users. We'd be happy to review and potentially incorporate your approach into the codebase.
